File: skills/markdown-deck/scripts/pack-repo.mjs
#!/usr/bin/env node
// SPDX-License-Identifier: Apache-2.0
/**
 * Assemble this skill into a standalone, publishable git repository.
 *
 * The installed shape and the published shape differ on purpose:
 *
 *   installed   .agents/skills/markdown-deck/SKILL.md      (name must match parent dir)
 *   published   markdown-deck/skills/markdown-deck/SKILL.md
 *
 * Nesting the payload under skills/ keeps repo scaffolding (plugin manifests, changelog,
 * CI) out of the directory that lands in every consumer's context window, and it is the
 * layout the install tooling expects: `gh skill install owner/repo markdown-deck`,
 * `npx skills add owner/repo/skills/markdown-deck`, and
 * `gemini skills install <url> --path skills/markdown-deck` all resolve it.
 *
 * Usage:
 *   node scripts/pack-repo.mjs --out ../../../../markdown-deck --owner patternode
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const SKILL_DIR = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const NAME = path.basename(SKILL_DIR);

const args = process.argv.slice(2);
const argOf = (k, d) => { const i = args.indexOf(k); return i >= 0 ? args[i + 1] : d; };
const outRoot = path.resolve(argOf('--out', path.join(SKILL_DIR, '..', '..', '..', '.tmp', NAME)));
const owner = argOf('--owner', 'OWNER');
const repoUrl = argOf('--repo', `https://github.com/${owner}/${NAME}`);

const pkg = JSON.parse(fs.readFileSync(path.join(SKILL_DIR, 'package.json'), 'utf8'));
const version = pkg.version;

// Files that belong to the repo, not to the portable skill payload.
const REPO_ONLY = new Set(['README.md', 'CHANGELOG.md', '.gitignore', '.gitattributes']);
const NEVER = new Set(['node_modules', '.git', '.tmp', 'dist']);

function copyTree(from, to, filter = () => true) {
  for (const e of fs.readdirSync(from, { withFileTypes: true })) {
    if (NEVER.has(e.name)) continue;
    const src = path.join(from, e.name);
    const rel = path.relative(SKILL_DIR, src).split(path.sep).join('/');
    if (!filter(rel, e)) continue;
    const dst = path.join(to, e.name);
    if (e.isDirectory()) { fs.mkdirSync(dst, { recursive: true }); copyTree(src, dst, filter); }
    else fs.copyFileSync(src, dst);
  }
}

fs.rmSync(outRoot, { recursive: true, force: true });
const payload = path.join(outRoot, 'skills', NAME);
fs.mkdirSync(payload, { recursive: true });

// 1. The portable payload. LICENSE stays inside it so a copied directory is legally
//    self-sufficient, which is what Anthropic's own published skills do.
copyTree(SKILL_DIR, payload, (rel) => !REPO_ONLY.has(rel) && !rel.startsWith('scripts/pack-repo'));

// 2. Repo-level files.
for (const f of REPO_ONLY) {
  const src = path.join(SKILL_DIR, f);
  if (fs.existsSync(src)) fs.copyFileSync(src, path.join(outRoot, f));
}
fs.copyFileSync(path.join(SKILL_DIR, 'LICENSE'), path.join(outRoot, 'LICENSE'));
fs.mkdirSync(path.join(outRoot, 'LICENSES'), { recursive: true });
fs.copyFileSync(path.join(SKILL_DIR, 'LICENSE'), path.join(outRoot, 'LICENSES', 'Apache-2.0.txt'));

// 3. Claude Code plugin + marketplace. This is the only distribution route that resolves
//    skill dependencies, so it is worth shipping even for a skill that has none today.
const pluginDir = path.join(outRoot, '.claude-plugin');
fs.mkdirSync(pluginDir, { recursive: true });
fs.writeFileSync(path.join(pluginDir, 'plugin.json'), JSON.stringify({
  name: NAME,
  displayName: 'Markdown Deck',
  version,
  description: pkg.description,
  author: { name: owner, url: `https://github.com/${owner}` },
  repository: repoUrl,
  license: pkg.license,
  keywords: pkg.keywords,
  skills: './skills/',
  dependencies: [],
}, null, 2) + '\n');
fs.writeFileSync(path.join(pluginDir, 'marketplace.json'), JSON.stringify({
  name: `${owner}-skills`,
  owner: { name: owner },
  plugins: [{
    name: NAME,
    source: './',
    description: pkg.description,
    version,
    category: 'documentation',
    tags: pkg.keywords,
  }],
}, null, 2) + '\n');

const count = (d) => fs.readdirSync(d, { withFileTypes: true })
  .reduce((n, e) => n + (NEVER.has(e.name) ? 0 : e.isDirectory() ? count(path.join(d, e.name)) : 1), 0);

console.log(`  packed ${NAME} v${version} -> ${outRoot}`);
console.log(`  ${count(outRoot)} files; payload at skills/${NAME}/`);
console.log('');
console.log('  next:');
console.log(`    cd ${outRoot}`);
console.log('    git init && git add -A && git commit -m "markdown-deck ' + version + '"');
console.log(`    git remote add origin ${repoUrl}.git && git push -u origin main`);
console.log(`    git tag v${version} && git push --tags`);

File: skills/model/scripts/pack-repo.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""Assemble this skill into a standalone, publishable git repository.

The installed shape and the published shape differ on purpose:

    installed   .agents/skills/model/SKILL.md      (name must match the parent directory)
    published   model/skills/model/SKILL.md

Nesting the payload under skills/ keeps repo scaffolding out of the directory that lands
in every consumer's context window, and it is the layout the install tooling expects:
`gh skill install owner/repo model`, `npx skills add owner/repo/skills/model` and
`gemini skills install <url> --path skills/model` all resolve it.

Usage:
  python scripts/pack-repo.py --owner patternode --out ../../../../model
"""
from __future__ import annotations

import argparse
import json
import os
import shutil

HERE = os.path.dirname(os.path.abspath(__file__))
SKILL_DIR = os.path.dirname(HERE)
NAME = os.path.basename(SKILL_DIR)

# Belongs to the repository, not to the portable payload.
REPO_ONLY = {"README.md", "CHANGELOG.md", ".gitignore", ".gitattributes"}
NEVER = {"__pycache__", ".git", ".tmp", "dist", ".pytest_cache"}


def version() -> str:
    init = os.path.join(SKILL_DIR, "src", NAME, "__init__.py")
    for line in open(init, encoding="utf-8"):
        if line.startswith("__version__"):
            return line.split("=", 1)[1].strip().strip('"').strip("'")
    return "0.0.0"


def copy_tree(src, dst, skip_rel=()):
    for root, dirs, files in os.walk(src):
        dirs[:] = [d for d in dirs if d not in NEVER]
        for f in files:
            full = os.path.join(root, f)
            rel = os.path.relpath(full, src).replace(os.sep, "/")
            if rel in REPO_ONLY or rel in skip_rel:
                continue
            out = os.path.join(dst, rel)
            os.makedirs(os.path.dirname(out), exist_ok=True)
            shutil.copy2(full, out)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--out", default=os.path.join(SKILL_DIR, ".tmp", "packed"))
    ap.add_argument("--owner", default="OWNER")
    ap.add_argument("--repo")
    a = ap.parse_args()

    v = version()
    out_root = os.path.abspath(a.out)
    repo_url = a.repo or f"https://github.com/{a.owner}/{NAME}"

    if os.path.exists(out_root):
        shutil.rmtree(out_root)
    payload = os.path.join(out_root, "skills", NAME)
    os.makedirs(payload, exist_ok=True)

    # 1. The portable payload. LICENSE stays inside it so a copied directory is
    #    legally self-sufficient, which is what published skills do.
    copy_tree(SKILL_DIR, payload, skip_rel=("scripts/pack-repo.py",))

    # 2. Repo-level files.
    for f in REPO_ONLY:
        src = os.path.join(SKILL_DIR, f)
        if os.path.exists(src):
            shutil.copy2(src, os.path.join(out_root, f))
    shutil.copy2(os.path.join(SKILL_DIR, "LICENSE"), os.path.join(out_root, "LICENSE"))
    os.makedirs(os.path.join(out_root, "LICENSES"), exist_ok=True)
    shutil.copy2(os.path.join(SKILL_DIR, "LICENSE"),
                 os.path.join(out_root, "LICENSES", "Apache-2.0.txt"))

    # 3. Claude Code plugin and marketplace. The only distribution route that resolves
    #    dependencies between skills, so it is worth shipping even with none declared.
    pdir = os.path.join(out_root, ".claude-plugin")
    os.makedirs(pdir, exist_ok=True)
    desc = ("Treat a diagram and a document as two views of one model of boxes and lines: "
            "extract, emit, validate and render across Markdown, draw.io, JSON and YAML.")
    write_json(os.path.join(pdir, "plugin.json"), {
        "name": NAME,
        "displayName": "Model",
        "version": v,
        "description": desc,
        "author": {"name": a.owner, "url": f"https://github.com/{a.owner}"},
        "repository": repo_url,
        "license": "Apache-2.0",
        "keywords": ["drawio", "markdown", "model", "diagram"],
        "skills": "./skills/",
        "dependencies": [],
    })
    write_json(os.path.join(pdir, "marketplace.json"), {
        "name": f"{a.owner}-skills",
        "owner": {"name": a.owner},
        "plugins": [{
            "name": NAME, "source": "./", "description": desc, "version": v,
            "category": "documentation", "tags": ["drawio", "markdown", "model"],
        }],
    })

    n = sum(len(f) for _r, _d, f in os.walk(out_root))
    print(f"  packed {NAME} v{v} -> {out_root}")
    print(f"  {n} files; payload at skills/{NAME}/")
    print("")
    print("  next:")
    print(f"    cd {out_root}")
    print(f'    git init && git add -A && git commit -m "{NAME} {v}"')
    print(f"    git remote add origin {repo_url}.git && git push -u origin main")
    print(f"    git tag v{v} && git push --tags")


def write_json(path, obj):
    with open(path, "w", encoding="utf-8", newline="\n") as fh:
        json.dump(obj, fh, indent=2, ensure_ascii=False)
        fh.write("\n")


if __name__ == "__main__":
    main()
